diff --git a/src/tracker-extract/tracker-extract-libav.c b/src/tracker-extract/tracker-extract-libav.c
index f53371574d41285c52c5dab1301e8eb729e315e2..bac4456005dd9a987b60700fc4c36e64c4ff8158 100644
--- a/src/tracker-extract/tracker-extract-libav.c
+++ b/src/tracker-extract/tracker-extract-libav.c
@@ -101,8 +101,8 @@ tracker_extract_get_metadata (TrackerExtractInfo  *info,
 		if (audio_stream->codecpar->sample_rate > 0) {
 			tracker_resource_set_int64 (metadata, "nfo:sampleRate", audio_stream->codecpar->sample_rate);
 		}
-		if (audio_stream->codecpar->channels > 0) {
-			tracker_resource_set_int64 (metadata, "nfo:channels", audio_stream->codecpar->channels);
+		if (audio_stream->codecpar->ch_layout.nb_channels > 0) {
+			tracker_resource_set_int64 (metadata, "nfo:channels", audio_stream->codecpar->ch_layout.nb_channels);
 		}
 	}
 
