diff -ru retro-plugins-0.8.1.old/plugins/game-boy/Makefile.in retro-plugins-0.8.1/plugins/game-boy/Makefile.in
--- retro-plugins-0.8.1.old/plugins/game-boy/Makefile.in	2016-09-13 09:19:18.000000000 +0300
+++ retro-plugins-0.8.1/plugins/game-boy/Makefile.in	2016-09-22 02:34:22.765473157 +0300
@@ -1,13 +1,13 @@
 SYSTEM_NAME = game-boy
 SUBMODULE_DIR = gambatte-libretro
 
-OUT = $(srcdir)/$(SUBMODULE_DIR)/libgambatte/gambatte_libretro.so
+OUT = $(srcdir)/$(SUBMODULE_DIR)/gambatte_libretro.so
 
 -include ../submodule-plugin.mk
 
 $(OUT):
-	make -C $(srcdir)/$(SUBMODULE_DIR)/libgambatte -f Makefile.libretro
+	make -C $(srcdir)/$(SUBMODULE_DIR) -f Makefile.libretro
 
 submodule-clean:
-	make -C $(srcdir)/$(SUBMODULE_DIR)/libgambatte -f Makefile.libretro clean
+	make -C $(srcdir)/$(SUBMODULE_DIR) -f Makefile.libretro clean
 
