diff -ru vte-0.48.0.old/bindings/vala/Makefile.am vte-0.48.0/bindings/vala/Makefile.am
--- vte-0.48.0.old/bindings/vala/Makefile.am	2017-03-20 23:56:32.000000000 +0300
+++ vte-0.48.0/bindings/vala/Makefile.am	2017-03-21 13:21:58.629525835 +0300
@@ -33,6 +33,7 @@
 vte_@VTE_API_MAJOR_VERSION@_@VTE_API_MINOR_VERSION@_SOURCES = \
 	app.vala \
 	config.vapi \
+	vte-$(VTE_API_VERSION).vapi \
 	$(NULL)
 
 nodist_vte_@VTE_API_MAJOR_VERSION@_@VTE_API_MINOR_VERSION@_SOURCES = \
