diff --git a/src/liblatexila/latexila-post-processor-latexmk.c b/src/liblatexila/latexila-post-processor-latexmk.c
index cae1b86fc7c131addaa2b6803e73e46ac5e63b3a..c053af7795d3d8ce93af6f7f480d27e0e60c0de0 100644
--- a/src/liblatexila/latexila-post-processor-latexmk.c
+++ b/src/liblatexila/latexila-post-processor-latexmk.c
@@ -241,7 +241,8 @@ fetch_sub_title (LatexilaPostProcessorLatexmk *pp,
 		rule = g_match_info_fetch_named (match_info, "rule");
 
 		pp->priv->last_rule_is_latex_rule = (g_strcmp0 (rule, "latex") == 0 ||
-						     g_strcmp0 (rule, "pdflatex") == 0);
+						     g_strcmp0 (rule, "pdflatex") == 0 ||
+						     g_strcmp0 (rule, "xelatex") == 0);
 
 		if (pp->priv->last_rule_is_latex_rule)
 		{
