diff -ru gnome-documents-3.18.1.old/data/Makefile.am gnome-documents-3.18.1/data/Makefile.am
--- gnome-documents-3.18.1.old/data/Makefile.am	2015-10-16 15:04:54.943922109 +0300
+++ gnome-documents-3.18.1/data/Makefile.am	2015-10-16 15:11:22.667433567 +0300
@@ -60,8 +60,7 @@
     $(desktop_DATA)
 
 install-exec-hook:
-	-rm -f $(appdir)/org.gnome.Books.data.gresource
-	$(LN_S) $(appdir)/org.gnome.Documents.data.gresource $(DESTDIR)$(appdir)/org.gnome.Books.data.gresource
+	$(LN_S) -f $(appdir)/org.gnome.Documents.data.gresource $(DESTDIR)$(appdir)/org.gnome.Books.data.gresource
 uninstall-hook:
 	-rm -f $(appdir)/org.gnome.Books.data.gresource
 
diff -ru gnome-documents-3.18.1.old/data/Makefile.in gnome-documents-3.18.1/data/Makefile.in
--- gnome-documents-3.18.1.old/data/Makefile.in	2015-10-16 15:04:54.946922086 +0300
+++ gnome-documents-3.18.1/data/Makefile.in	2015-10-16 15:11:10.857510235 +0300
@@ -843,8 +843,7 @@
 @GSETTINGS_RULES@
 
 install-exec-hook:
-	-rm -f $(appdir)/org.gnome.Books.data.gresource
-	$(LN_S) $(appdir)/org.gnome.Documents.data.gresource $(DESTDIR)$(appdir)/org.gnome.Books.data.gresource
+	$(LN_S) -f $(appdir)/org.gnome.Documents.data.gresource $(DESTDIR)$(appdir)/org.gnome.Books.data.gresource
 uninstall-hook:
 	-rm -f $(appdir)/org.gnome.Books.data.gresource
 
