diff -ru gnome-system-monitor-3.25.92.old/src/prettytable.cpp gnome-system-monitor-3.25.92/src/prettytable.cpp
--- gnome-system-monitor-3.25.92.old/src/prettytable.cpp	2017-06-19 21:45:16.000000000 +0300
+++ gnome-system-monitor-3.25.92/src/prettytable.cpp	2017-09-07 01:28:58.250247507 +0300
@@ -104,9 +104,10 @@
 {
   /* If process already exists then set the icon. Otherwise put into hash
   ** table to be added later */
+  try {  
     if (ProcInfo* info = GsmApplication::get()->processes.find(pid))
     {
-      info.set_icon(icon);
+      info->set_icon(icon);
       // move the ref to the map
       this->apps[pid] = icon;
       procman_debug("WNCK OK for %u", unsigned(pid));
