diff --git a/configure.ac b/configure.ac
index 9c1fdb4b..7fb1da9d 100644
--- a/configure.ac
+++ b/configure.ac
@@ -176,6 +176,12 @@ AC_ARG_WITH(dmconfdir,
             dmconfdir=${withval}, dmconfdir=${sysconfdir}/dm)
 AC_SUBST(dmconfdir)
 
+AC_ARG_WITH([udevrulesdir],
+            AS_HELP_STRING([--with-udevrulesdir=DIR], [Directory for udev rules files]),
+            [],
+            [with_udevrulesdir=$($PKG_CONFIG --variable=udevdir udev)/rules.d])
+AC_SUBST([udevrulesdir], [$with_udevrulesdir])
+
 dnl ---------------------------------------------------------------------------
 dnl - Configure arguments
 dnl ---------------------------------------------------------------------------
diff --git a/data/Makefile.am b/data/Makefile.am
index a47e7900..88270594 100644
--- a/data/Makefile.am
+++ b/data/Makefile.am
@@ -158,7 +158,6 @@ if ENABLE_ARCH_PAM_CONFIG
 pam_files = $(pam_arch_files)
 endif
 
-udevrulesdir = $(prefix)/lib/udev/rules.d
 udevrules_DATA = 61-gdm.rules
 
 EXTRA_DIST +=			\
