diff --git a/src/core/display.c b/src/core/display.c
index a66ddb1c6..1866f9be0 100644
--- a/src/core/display.c
+++ b/src/core/display.c
@@ -51,6 +51,7 @@
 #include <meta/meta-enum-types.h>
 #include "meta-idle-monitor-dbus.h"
 #include "meta-cursor-tracker-private.h"
+#include "meta-cursor-sprite-xcursor.h"
 #include <meta/meta-backend.h>
 #include "backends/meta-logical-monitor.h"
 #include "backends/native/meta-backend-native.h"
