diff --unified --new-file --recursive --show-c-function mutter-3.34.0/src/compositor/compositor.c mutter-3.34.0-fix/src/compositor/compositor.c
--- mutter-3.34.0/src/compositor/compositor.c	2019-10-05 09:39:37.210269887 +0200
+++ mutter-3.34.0-fix/src/compositor/compositor.c	2019-10-05 09:47:11.710916909 +0200
@@ -63,7 +63,6 @@
 #include "clutter/clutter-mutter.h"
 #include "cogl/cogl-trace.h"
 #include "compositor/meta-window-actor-x11.h"
-#include "compositor/meta-window-actor-wayland.h"
 #include "compositor/meta-window-actor-private.h"
 #include "compositor/meta-window-group-private.h"
 #include "core/core.h"
@@ -84,6 +83,7 @@
 
 #ifdef HAVE_WAYLAND
 #include "wayland/meta-wayland-private.h"
+#include "compositor/meta-window-actor-wayland.h"
 #endif
 
 enum
