diff -ru ../mutter-3.25.91.old/src/backends/native/meta-launcher.c ./src/backends/native/meta-launcher.c
--- ../mutter-3.25.91.old/src/backends/native/meta-launcher.c	2017-09-06 15:24:45.757337209 +0300
+++ ./src/backends/native/meta-launcher.c	2017-09-06 15:24:58.897070346 +0300
@@ -49,9 +49,9 @@
 
 #define DRM_CARD_UDEV_DEVICE_TYPE "drm_minor"
 
-G_DEFINE_AUTOPTR_CLEANUP_FUNC(GUdevDevice, g_object_unref)
-G_DEFINE_AUTOPTR_CLEANUP_FUNC(GUdevClient, g_object_unref)
-G_DEFINE_AUTOPTR_CLEANUP_FUNC(GUdevEnumerator, g_object_unref)
+//G_DEFINE_AUTOPTR_CLEANUP_FUNC(GUdevDevice, g_object_unref)
+//G_DEFINE_AUTOPTR_CLEANUP_FUNC(GUdevClient, g_object_unref)
+//G_DEFINE_AUTOPTR_CLEANUP_FUNC(GUdevEnumerator, g_object_unref)
 
 struct _MetaLauncher
 {
