diff -ru epiphany-3.26.0.old/tests/ephy-file-helpers-test.c epiphany-3.26.0/tests/ephy-file-helpers-test.c
--- epiphany-3.26.0.old/tests/ephy-file-helpers-test.c	2017-09-09 07:16:25.000000000 +0300
+++ epiphany-3.26.0/tests/ephy-file-helpers-test.c	2017-09-10 12:50:37.332338664 +0300
@@ -85,12 +85,7 @@
 
     /* Cleanup dir left behind. */
     if (keep_dir) {
-      /* As a safety measure, only try recursive delete on paths
-       * prefixed with /tmp. */
-      if (g_str_has_prefix (tmp_dir, "/tmp"))
         g_assert (ephy_file_delete_dir_recursively (tmp_dir, NULL));
-      else
-        g_warning ("INIT: dangerous path returned as tmp_dir: %s", tmp_dir);
     }
 
     g_free (tmp_dir);
