diff -ru gnome-control-center-3.18.0.old/configure.ac gnome-control-center-3.18.0/configure.ac
--- gnome-control-center-3.18.0.old/configure.ac	2015-09-23 01:44:20.710320437 +0300
+++ gnome-control-center-3.18.0/configure.ac	2015-09-23 01:46:29.973294724 +0300
@@ -526,22 +526,6 @@
         AC_MSG_RESULT(no)
 fi
 
-case $host_os in
-  linux*)
-    if test "$host_cpu" = s390 -o "$host_cpu" = s390x; then
-      # Nothing
-      echo "USB-related panels disabled on S390"
-    else
-      if test "x$have_networkmanager" != "xyes" -o \
-              "x$have_bluetooth" != "xyes" -o \
-              "x$enable_cups" != "xyes" -o \
-              "x$have_wacom" != "xyes"; then
-        AC_MSG_ERROR([The Network, Bluetooth, Printers and Wacom panels are not optional on Linux systems])
-     fi
-    fi
-    ;;
-esac
-
 AC_CONFIG_FILES([
 Makefile
 panels/Makefile
