From 55fe87357c27c75de139c63e92e9b599f46699e1 Mon Sep 17 00:00:00 2001
From: Mart Raudsepp <leio@gentoo.org>
Date: Sun, 8 Sep 2019 18:07:12 +0300
Subject: [PATCH] build: Allow NM optional on Linux

---
 meson.build | 1 -
 1 file changed, 1 deletion(-)

diff --git a/meson.build b/meson.build
index e1ef82e4..d71a40a8 100644
--- a/meson.build
+++ b/meson.build
@@ -198,7 +198,6 @@ endif
 
 # Sharing plugin
 enable_network_manager = get_option('network_manager')
-assert(enable_network_manager or not host_is_linux, 'NetworkManager support is not optional on Linux platforms')
 if enable_network_manager
   # network manager
   libnm_dep = dependency('libnm', version: '>= 1.0')
-- 
2.20.1

