From 44fc7f067fcb43da7c287a8a545137ce5d601b07 Mon Sep 17 00:00:00 2001
From: Poncho <poncho@spahan.ch>
Date: Sat, 17 Mar 2018 11:13:57 +0100
Subject: [PATCH 5/5] fix build without wayland

---
 plugins/common/gsd-device-manager-udev.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/plugins/common/gsd-device-manager-udev.c b/plugins/common/gsd-device-manager-udev.c
index 37581aae..9559cba5 100644
--- a/plugins/common/gsd-device-manager-udev.c
+++ b/plugins/common/gsd-device-manager-udev.c
@@ -24,7 +24,10 @@
 #include <string.h>
 #include <gudev/gudev.h>
 
+#ifdef GDK_WINDOWING_WAYLAND
 #include <gdk/gdkwayland.h>
+#endif
+
 #include "gsd-device-manager-udev.h"
 
 struct _GsdUdevDeviceManager
@@ -244,6 +247,7 @@ static GsdDevice *
 gsd_udev_device_manager_lookup_device (GsdDeviceManager *manager,
 				       GdkDevice	*gdk_device)
 {
+#ifdef GDK_WINDOWING_WAYLAND
 	const gchar *node_path;
 	GHashTableIter iter;
 	GsdDevice *device;
@@ -260,6 +264,7 @@ gsd_udev_device_manager_lookup_device (GsdDeviceManager *manager,
 			return device;
 		}
 	}
+#endif
 
 	return NULL;
 }
-- 
2.16.1

