diff --git a/podcasts-gtk/src/meson.build b/podcasts-gtk/src/meson.build
index a992b6a..7b55e9b 100644
--- a/podcasts-gtk/src/meson.build
+++ b/podcasts-gtk/src/meson.build
@@ -28,8 +28,6 @@ else
   message('Building in debug mode')
 endif
 
-cargo_env = [ 'CARGO_HOME=' + meson.project_build_root() / 'cargo-home' ]
-
 cargo_release = custom_target('cargo-build',
                         build_by_default: true,
                         build_always_stale: true,
@@ -39,7 +37,6 @@ cargo_release = custom_target('cargo-build',
                         console: true,
                         depends: podcasts_resources,
                         command: ['env',
-                                  cargo_env,
                                   cargo, 'build',
                                   cargo_options,
                                   '&&',
