diff --git a/src/user-classify.c b/src/user-classify.c
index 9db1c9a..1e17e78 100644
--- a/src/user-classify.c
+++ b/src/user-classify.c
@@ -52,6 +52,48 @@ static const char *default_excludes[] = {
         "at",
         "gdm",
         "gnome-initial-setup"
+        /* Additional Gentoo system users with non-trivial login shell */
+        "amanda",
+        "backuppc",
+        "drqueue",
+        "firebird",
+        "flexlm",
+        "foldingathome",
+        "geneweb",
+        "git",
+        "gnump3d",
+        "hacluster",
+        "hg",
+        "hsqldb",
+        "infinote",
+        "foldingathome",
+        "geneweb",
+        "git",
+        "gnump3d",
+        "hacluster",
+        "hg",
+        "hsqldb",
+        "infinote",
+        "jffnms",
+        "klive",
+        "mailman",
+        "mpd",
+        "mythtv",
+        "nagios",
+        "nx",
+        "oneadmin",
+        "openvpn",
+        "p2p",
+        "phxd",
+        "resin",
+        "rplayd",
+        "scponly",
+        "secoff",
+        "tinyproxy",
+        "ttrssd",
+        "vboxguest",
+        "vdr",
+        "vdradmin",
 };
 
 static gboolean
