From f1f551bf5d4d9fcf32504041d02423d30660bd94 Mon Sep 17 00:00:00 2001
From: Poncho <poncho@spahan.ch>
Date: Mon, 10 Sep 2018 11:17:29 +0200
Subject: [PATCH 2/2] meson.build: disable bubblewrap

---
 meson.build | 1 -
 1 file changed, 1 deletion(-)

diff --git a/meson.build b/meson.build
index 661ba7850..de1f4a391 100644
--- a/meson.build
+++ b/meson.build
@@ -175,7 +175,6 @@ conf.set_quoted('VERSION', '@0@-@VCS_TAG@'.format(meson.project_version()))
 ###################################################
 
 conf.set('ENABLE_SECCOMP', seccomp.found())
-conf.set('HAVE_BWRAP', seccomp.found())
 
 conf.set('ENABLE_PACKAGEKIT', get_option('packagekit'))
 conf.set('ENABLE_PROFILING', get_option('profiling'))
-- 
2.16.4

