diff -ru nautilus-3.16.0.old/src/nautilus-application.c nautilus-3.16.0/src/nautilus-application.c
--- nautilus-3.16.0.old/src/nautilus-application.c	2015-03-24 19:06:50.000000000 +0300
+++ nautilus-3.16.0/src/nautilus-application.c	2015-03-27 01:48:26.063995817 +0300
@@ -761,7 +761,7 @@
 			file = g_file_new_for_commandline_arg (remaining[idx]);
 			g_ptr_array_add (file_array, file);
 		}
-	} else if (g_variant_dict_contains (options, "new-window")) {
+    if (g_variant_dict_contains (options, "no-default-window")) {
 		file = g_file_new_for_path (g_get_home_dir ());
 		g_ptr_array_add (file_array, file);
 	} else {
