diff -ru gnome-tweak-tool-3.26.0.old/gtweak/tweaks/tweak_group_shell_extensions.py gnome-tweak-tool-3.26.0/gtweak/tweaks/tweak_group_shell_extensions.py
--- gnome-tweak-tool-3.26.0.old/gtweak/tweaks/tweak_group_shell_extensions.py	2017-09-11 21:38:20.000000000 +0300
+++ gnome-tweak-tool-3.26.0/gtweak/tweaks/tweak_group_shell_extensions.py	2017-09-20 08:58:59.850858322 +0300
@@ -85,8 +85,7 @@
         self._shell = shell
         state = ext.get("state")
         uuid = ext["uuid"]
-        uuid = uuid.replace('@', '_')
-        self._app_id = "user/*/extensions-web/shell-extension/" + uuid + "/*"
+        self._app_id = "user/*/extensions-web/shell-extension/" + uuid.replace('@', '_') + "/*"
 
         shell._settings.bind("disable-user-extensions", self,
                              "sensitive", Gio.SettingsBindFlags.INVERT_BOOLEAN)
