diff --git a/configure.ac b/configure.ac
index 118d04f3..085d6a64 100644
--- a/configure.ac
+++ b/configure.ac
@@ -167,7 +167,6 @@ case $host_os in
         ;;
     esac
     AC_DEFINE_UNQUOTED(_GNU_SOURCE, 1, [Define to include GNU extensions])
-    AC_DEFINE_UNQUOTED(HAVE_BWRAP, 1, [Define to 1 if Bubblewrap support is available])
     AC_DEFINE_UNQUOTED(INSTALL_PREFIX, "$prefix", [Path to library install prefix])
     ;;
 esac
